### ISSUE TITLE
Building on mingw-w64 (ANSI C portability)

### ISSUE THREAD

**Issue by jeroenooms** [WizardMac/ReadStat#13]
I'm trying to build this library with `mingw-w64` (the windows build of gcc) but there are some non portable pieces. First attempt:

```bash
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"     -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_convert.c -o readstat_convert.o
readstat_bits.c: In function 'byteswap_float':
readstat_bits.c:59:5: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
readstat_bits.c:60:5: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
readstat_bits.c: In function 'byteswap_double':
readstat_bits.c:64:5: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
readstat_bits.c:65:5: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"     -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_dta.c -o readstat_dta.o
readstat_convert.c: In function 'readstat_convert':
readstat_convert.c:23:9: warning: passing argument 2 of 'libiconv' from incompatible pointer type [enabled by default]
C:/PROGRA~1/R/R-31~1.2/include/iconv.h:53:8: note: expected 'const char **' but argument is of type 'char **'
readstat_dta.c:6:21: fatal error: sys/uio.h: No such file or directory
compilation terminated.
```

I tried using:

```c
#ifndef WIN32
#include <sys/uio.h>
#endif
```

But that makes things worse:

```bash
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"     -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_por_parse.c -o readstat_por_parse.o
src/readstat_por_parse.rl: In function 'readstat_por_parse_double':
src/readstat_por_parse.rl:21:5: error: unknown type name 'u_char'
src/readstat_por_parse.rl:23:5: error: unknown type name 'u_char'
src/readstat_por_parse.rl:23:24: error: 'u_char' undeclared (first use in this function)
src/readstat_por_parse.rl:23:24: note: each undeclared identifier is reported only once for each function it appears in
src/readstat_por_parse.rl:23:32: error: expected expression before ')' token
src/readstat_por_parse.rl:24:18: error: expected '=', ',', ';', 'asm' or '__attribute__' before '*' token
src/readstat_por_parse.rl:24:34: error: expected expression before ')' token
src/readstat_por_parse.c:353:9: warning: comparison of distinct pointer types lacks a cast [enabled by default]
src/readstat_por_parse.rl:69:9: warning: implicit declaration of function 'dprintf' [-Wimplicit-function-declaration]
src/readstat_por_parse.rl:69:84: error: expected expression before ')' token
src/readstat_por_parse.rl:76:32: error: expected expression before ')' token
make: *** [readstat_por_parse.o] Error 1
```

I think the problem is that `u_char` is not part of the ANSI C standard. The typical alternative  is to use `<stdint.h>` for the `uint8_t` type, or `unsigned char`.

I'm no expert on this so unfortunately I can't further narrow down the problem or provide a PR. Some other data intense libraries that I have worked with that do build on windows are [yajl](https://github.com/lloyd/yajl/), [libgit2](https://github.com/libgit2/libgit2) and [mongo-c-driver](https://github.com/mongodb/mongo-c-driver), perhaps we can have a peek.

What I can do is help with testing, or provide a compiler for you to test.

**Comment by evanmiller** [comment]
Looks like <sys/uio.h> is unneeded. Try this:

https://github.com/WizardMac/ReadStat/commit/a49c7415503b9068b057ec0389701edec7179a87

**Comment by evanmiller** [comment]
...the `u_char` errors are a separate problem. Working on a fix.

**Comment by evanmiller** [comment]
This commit should fix the `u_char` errors:

https://github.com/WizardMac/ReadStat/commit/a45406bcc426ba0fd96d4a8455143df743d49306

**Comment by evanmiller** [comment]
This commit addresses the `dprintf` warnings:

https://github.com/WizardMac/ReadStat/commit/eb6b293c9f55c52cc879ebd1c440d2060ba1ed27

**Comment by jeroenooms** [comment]
Hmmm im getting an error on osx now, not sure if that is a problem on my end though:

```bash
clang -I/Library/Frameworks/R.framework/Resources/include -DNDEBUG  -I/usr/local/include -I/usr/local/include/freetype2 -I/opt/X11/include -I"/Library/Frameworks/R.framework/Versions/3.1/Resources/library/Rcpp/include"   -fPIC  -Wall -mtune=core2 -g -O2  -c readstat_rdata.c -o readstat_rdata.o
src/readstat_por_parse.c:11:18: warning: unused variable 'por_field_parse_first_final' [-Wunused-const-variable]
static const int por_field_parse_first_final = 12;
                 ^
src/readstat_por_parse.c:12:18: warning: unused variable 'por_field_parse_error' [-Wunused-const-variable]
static const int por_field_parse_error = 0;
                 ^
src/readstat_por_parse.c:14:18: warning: unused variable 'por_field_parse_en_main' [-Wunused-const-variable]
static const int por_field_parse_en_main = 1;
                 ^
3 warnings generated.
readstat_rdata.c:16:10: fatal error: 'lzma.h' file not found
#include <lzma.h>
         ^
1 error generated.
```

**Comment by evanmiller** [comment]
This is a bug. lzma.dylib should be present on OS X 10.9 and 10.10, but the headers need to be installed manually.

**Comment by dennisfisher** [comment]
Evan

I am not sure that is true.  I just searched my system for dylib files — I found many but not that one.  Interestingly, I can run Hadley’s code on that same machine without the lzma.h error.  But, I encounter that error on a different OS X machine.  I have a have a vague recollection of installing lzma on one or more machines.

Regardless, you refer to this as a “bug … but the headers need to be installed manually”.  Is that something at the user end?

Dennis


Dennis Fisher MD
P < (The "P Less Than" Company)
Phone: 1-866-PLessThan (1-866-753-7784)
Fax: 1-866-PLessThan (1-866-753-7784)
www.PLessThan.com

**Comment by jeroenooms** [comment]
Some more warnings on windows:

```c
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"     -O2 -Wall  -std=gnu99 -mtune=core2 -c CKHashTable.c -o CKHashTable.o
* installing to library 'C:/Users/Jeroen/Documents/R/win-library/3.1'
* installing *source* package 'haven' ...
** libs
g++ -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"     -O2 -Wall  -mtune=core2 -c DfBuilder.cpp -o DfBuilder.o
g++ -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"     -O2 -Wall  -mtune=core2 -c RcppExports.cpp -o RcppExports.o
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"     -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_bits.c -o readstat_bits.o
readstat_bits.c: In function 'byteswap_float':
readstat_bits.c:59:5: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
readstat_bits.c:60:5: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
readstat_bits.c: In function 'byteswap_double':
readstat_bits.c:64:5: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
readstat_bits.c:65:5: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"     -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_convert.c -o readstat_convert.o
readstat_convert.c: In function 'readstat_convert':
readstat_convert.c:23:9: warning: passing argument 2 of 'libiconv' from incompatible pointer type [enabled by default]
C:/PROGRA~1/R/R-31~1.2/include/iconv.h:53:8: note: expected 'const char **' but argument is of type 'char **'
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"     -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_dta.c -o readstat_dta.o
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"     -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_error.c -o readstat_error.o
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"     -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_io.c -o readstat_io.o
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"     -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_por.c -o readstat_por.o
readstat_por.c: In function 'read_por_file_data':
readstat_por.c:549:21: warning: implicit declaration of function 'dprintf' [-Wimplicit-function-declaration]
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"     -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_por_parse.c -o readstat_por_parse.o
src/readstat_por_parse.rl: In function 'readstat_por_parse_double':
src/readstat_por_parse.rl:69:9: warning: implicit declaration of function 'dprintf' [-Wimplicit-function-declaration]
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"     -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_sas.c -o readstat_sas.o
readstat_sas.c: In function 'sas_read_header':
readstat_sas.c:204:20: warning: variable 'a2' set but not used [-Wunused-but-set-variable]
readstat_sas.c: In function 'sas_parse_page_pass1':
readstat_sas.c:791:23: warning: variable 'subheader_type' set but not used [-Wunused-but-set-variable]
readstat_sas.c:773:14: warning: variable 'page_type' set but not used [-Wunused-but-set-variable]
readstat_sas.c: In function 'sas_parse_page_pass2':
readstat_sas.c:864:27: warning: variable 'subheader_type' set but not used [-Wunused-but-set-variable]
readstat_sas.c: In function 'parse_sas7bdat':
readstat_sas.c:1032:9: warning: implicit declaration of function 'dprintf' [-Wimplicit-function-declaration]
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"     -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_sav.c -o readstat_sav.o
readstat_sav.c: In function 'sav_read_value_label_record':
readstat_sav.c:387:13: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
readstat_sav.c: In function 'handle_missing_double':
readstat_sav.c:451:5: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
readstat_sav.c:453:5: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
readstat_sav.c:455:5: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
readstat_sav.c: In function 'sav_read_data':
readstat_sav.c:540:29: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
readstat_sav.c:639:17: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
readstat_sav.c: In function 'sav_parse_machine_integer_info_record':
readstat_sav.c:691:13: warning: implicit declaration of function 'dprintf' [-Wimplicit-function-declaration]
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"     -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_sav_parse.c -o readstat_sav_parse.o
src/readstat_sav_parse.rl: In function 'sav_parse_long_variable_names_record':
src/readstat_sav_parse.rl:56:17: warning: passing argument 2 of 'libiconv' from incompatible pointer type [enabled by default]
C:/PROGRA~1/R/R-31~1.2/include/iconv.h:53:8: note: expected 'const char **' but argument is of type 'char **'
src/readstat_sav_parse.rl:75:17: warning: implicit declaration of function 'dprintf' [-Wimplicit-function-declaration]
src/readstat_sav_parse.rl: In function 'sav_parse_very_long_string_record':
src/readstat_sav_parse.rl:154:17: warning: passing argument 2 of 'libiconv' from incompatible pointer type [enabled by default]
C:/PROGRA~1/R/R-31~1.2/include/iconv.h:53:8: note: expected 'const char **' but argument is of type 'char **'
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"     -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_spss.c -o readstat_spss.o
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"     -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_value.c -o readstat_value.o
```

**Comment by evanmiller** [comment]
I think the only problematic warning is the `dprintf` stuff which the latest code should fix.

**Comment by evanmiller** [comment]
@dennisfisher: I believe Hadley's code works because it leaves out readstat_rdata.c, which is the only file that uses the lzma stuff.

Regarding header installation what I meant was that in order to work the user should install XZ Utils here:

http://tukaani.org/xz/

We need to make this optional somehow though.

**Comment by hadley** [comment]
I think the last piece is the iconv problems:

```R
gcc  -I"D:/RCompile/recent/R-3.1.2/include"            -I"d:/RCompile/CRANpkg/lib/3.1/Rcpp/include" -I"d:/RCompile/CRANpkg/lib/3.1/BH/include" -I"d:/Rcompile/CRANpkg/extralibs215/local215/include"     -O3 -Wall  -std=gnu99 -mtune=core2 -c readstat_sav_parse.c -o readstat_sav_parse.o
src/readstat_sav_parse.rl: In function 'sav_parse_long_variable_names_record':
src/readstat_sav_parse.rl:56:17: warning: passing argument 2 of 'libiconv' from incompatible pointer type [enabled by default]
D:/RCompile/recent/R-3.1.2/include/iconv.h:53:8: note: expected 'const char **' but argument is of type 'char **'
src/readstat_sav_parse.rl:109:17: warning: format '%ld' expects argument of type 'long int', but argument 4 has type 'int' [-Wformat]
src/readstat_sav_parse.rl: In function 'sav_parse_very_long_string_record':
src/readstat_sav_parse.rl:154:17: warning: passing argument 2 of 'libiconv' from incompatible pointer type [enabled by default]
D:/RCompile/recent/R-3.1.2/include/iconv.h:53:8: note: expected 'const char **' but argument is of type 'char **'
src/readstat_sav_parse.rl:204:9: warning: format '%ld' expects argument of type 'long int', but argument 3 has type 'int' [-Wformat]
src/readstat_sav_parse.rl:204:9: warning: format '%ld' expects argument of type 'long int', but argument 4 has type 'int' [-Wformat]
gcc  -I"D:/RCompile/recent/R-3.1.2/include"            -I"d:/RCompile/CRANpkg/lib/3.1/Rcpp/include" -I"d:/RCompile/CRANpkg/lib/3.1/BH/include" -I"d:/Rcompile/CRANpkg/extralibs215/local215/include"     -O3 -Wall  -std=gnu99 -mtune=core2 -c readstat_spss.c -o readstat_spss.o
gcc  -I"D:/RCompile/recent/R-3.1.2/include"            -I"d:/RCompile/CRANpkg/lib/3.1/Rcpp/include" -I"d:/RCompile/CRANpkg/lib/3.1/BH/include" -I"d:/Rcompile/CRANpkg/extralibs215/local215/include"     -O3 -Wall  -std=gnu99 -mtune=core2 -c readstat_value.c -o readstat_value.o
g++ -shared -s -static-libgcc -o haven.dll tmp.def CKHashTable.o DfBuilder.o RcppExports.o readstat_bits.o readstat_convert.o readstat_dta.o readstat_error.o readstat_io.o readstat_por.o readstat_por_parse.o readstat_rdata.o readstat_sas.o readstat_sav.o readstat_sav_parse.o readstat_spss.o readstat_value.o -Ld:/Rcompile/CRANpkg/extralibs215/local215/lib/i386 -Ld:/Rcompile/CRANpkg/extralibs215/local215/lib -LD:/RCompile/recent/R-3.1.2/bin/i386 -lR
readstat_convert.o:readstat_convert.c:(.text+0x42): undefined reference to `_imp__libiconv'
readstat_rdata.o:readstat_rdata.c:(.text+0x87): undefined reference to `lzma_code'
readstat_rdata.o:readstat_rdata.c:(.text+0x153): undefined reference to `inflate'
readstat_rdata.o:readstat_rdata.c:(.text+0x1cb9): undefined reference to `lzma_end'
readstat_rdata.o:readstat_rdata.c:(.text+0x1cd3): undefined reference to `inflateEnd'
readstat_rdata.o:readstat_rdata.c:(.text+0x1dd8): undefined reference to `lzma_end'
readstat_rdata.o:readstat_rdata.c:(.text+0x1df2): undefined reference to `inflateEnd'
readstat_rdata.o:readstat_rdata.c:(.text+0x1fe4): undefined reference to `inflateInit_'
readstat_rdata.o:readstat_rdata.c:(.text+0x203e): undefined reference to `lzma_stream_decoder'
readstat_sas.o:readstat_sas.c:(.text+0x10e5): undefined reference to `_imp__libiconv_close'
readstat_sas.o:readstat_sas.c:(.text+0x1158): undefined reference to `_imp__libiconv_open'
readstat_sas.o:readstat_sas.c:(.text+0x1e0c): undefined reference to `_imp__libiconv_open'
readstat_sas.o:readstat_sas.c:(.text+0x1e59): undefined reference to `_imp__libiconv_close'
readstat_sav.o:readstat_sav.c:(.text+0x135): undefined reference to `_imp__libiconv_close'
readstat_sav.o:readstat_sav.c:(.text+0x318): undefined reference to `_imp__libiconv_close'
readstat_sav.o:readstat_sav.c:(.text+0x5d8): undefined reference to `_imp__libiconv_open'
readstat_sav_parse.o:readstat_sav_parse.c:(.text+0x12c): undefined reference to `_imp__libiconv'
readstat_sav_parse.o:readstat_sav_parse.c:(.text+0x4bde): undefined reference to `_imp__libiconv'
collect2: ld returned 1 exit status
```

(This is from win-builder, a CRAN building service).

**Comment by evanmiller** [comment]
These iconv errors look identical to the ones Dennis reported earlier on Windows:

http://r.789695.n4.nabble.com/Failure-to-execute-R-CMD-SHLIB-successfully-td4702587.html

The other undefined references can be fixed by linking against libzma and libz -- or by excluding `readstat_rdata.c` from the compilation.

**Comment by dennisfisher** [comment]
Evan

In this case, my did the identical thing on 2 machines
	install_github("hadley/haven")
	require(“haven”)
	read_sas(FILENAME)
I did these steps seconds apart so it is unlikely that Hadley had removed readstat_rdata.c (it is still there in https://github.com/hadley/haven/tree/master/src)

It worked on one machine.  On the other, I got the lzma error.  

The difference must be that one machine has lzma installed, the other not.  I trust that you and Hadley can figure out how to deal with this.  It does make sense to exclude readstat_rdata.c from his installation since is would be recursive to add software to R that enables reading of R files!  However, it is still there in github

Dennis

 
Dennis Fisher MD
P < (The "P Less Than" Company)
Phone: 1-866-PLessThan (1-866-753-7784)
Fax: 1-866-PLessThan (1-866-753-7784)
www.PLessThan.com

**Comment by jeroenooms** [comment]
The iconv problems can be fixed by linking to `Riconv.dll` that is included with R on windows. I have created a PR for hadley to fix that. 

The package now builds, however it still gives a bunch of warnings. Perhaps these are harmless, but it would be great to get these fixed:

```c
==> Rcmd.exe INSTALL --preclean --no-multiarch --with-keep.source haven

* installing to library 'C:/Users/Jeroen/Documents/R/win-library/3.1'
* installing *source* package 'haven' ...
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"C:/Users/Jeroen/Documents/R/win-library/3.1/BH/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"     -O2 -Wall  -std=gnu99 -mtune=core2 -c CKHashTable.c -o CKHashTable.o
** libs
g++ -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"C:/Users/Jeroen/Documents/R/win-library/3.1/BH/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"     -O2 -Wall  -mtune=core2 -c DfBuilder.cpp -o DfBuilder.o
g++ -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"C:/Users/Jeroen/Documents/R/win-library/3.1/BH/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"     -O2 -Wall  -mtune=core2 -c RcppExports.cpp -o RcppExports.o
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"C:/Users/Jeroen/Documents/R/win-library/3.1/BH/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"     -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_bits.c -o readstat_bits.o
readstat_bits.c: In function 'byteswap_float':
readstat_bits.c:59:5: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
readstat_bits.c:60:5: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
readstat_bits.c: In function 'byteswap_double':
readstat_bits.c:64:5: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
readstat_bits.c:65:5: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"C:/Users/Jeroen/Documents/R/win-library/3.1/BH/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"     -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_convert.c -o readstat_convert.o
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"C:/Users/Jeroen/Documents/R/win-library/3.1/BH/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"     -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_dta.c -o readstat_dta.o
readstat_convert.c: In function 'readstat_convert':
readstat_convert.c:23:9: warning: passing argument 2 of 'libiconv' from incompatible pointer type [enabled by default]
C:/PROGRA~1/R/R-31~1.2/include/iconv.h:53:8: note: expected 'const char **' but argument is of type 'char **'
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"C:/Users/Jeroen/Documents/R/win-library/3.1/BH/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"     -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_error.c -o readstat_error.o
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"C:/Users/Jeroen/Documents/R/win-library/3.1/BH/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"     -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_io.c -o readstat_io.o
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"C:/Users/Jeroen/Documents/R/win-library/3.1/BH/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"     -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_por.c -o readstat_por.o
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"C:/Users/Jeroen/Documents/R/win-library/3.1/BH/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"     -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_por_parse.c -o readstat_por_parse.o
src/readstat_por_parse.rl: In function 'readstat_por_parse_double':
src/readstat_por_parse.rl:69:9: warning: format '%ld' expects argument of type 'long int', but argument 3 has type 'long long int' [-Wformat]
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"C:/Users/Jeroen/Documents/R/win-library/3.1/BH/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"     -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_sas.c -o readstat_sas.o
readstat_sas.c: In function 'sas_read_header':
readstat_sas.c:204:20: warning: variable 'a2' set but not used [-Wunused-but-set-variable]
readstat_sas.c: In function 'sas_parse_page_pass1':
readstat_sas.c:791:23: warning: variable 'subheader_type' set but not used [-Wunused-but-set-variable]
readstat_sas.c:773:14: warning: variable 'page_type' set but not used [-Wunused-but-set-variable]
readstat_sas.c: In function 'sas_parse_page_pass2':
readstat_sas.c:864:27: warning: variable 'subheader_type' set but not used [-Wunused-but-set-variable]
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"C:/Users/Jeroen/Documents/R/win-library/3.1/BH/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"     -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_sav.c -o readstat_sav.o
readstat_sav.c: In function 'sav_read_value_label_record':
readstat_sav.c:387:13: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
readstat_sav.c: In function 'handle_missing_double':
readstat_sav.c:451:5: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
readstat_sav.c:453:5: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
readstat_sav.c:455:5: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
readstat_sav.c: In function 'sav_read_data':
readstat_sav.c:540:29: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
readstat_sav.c:639:17: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"C:/Users/Jeroen/Documents/R/win-library/3.1/BH/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"     -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_sav_parse.c -o readstat_sav_parse.o
src/readstat_sav_parse.rl: In function 'sav_parse_long_variable_names_record':
src/readstat_sav_parse.rl:56:17: warning: passing argument 2 of 'libiconv' from incompatible pointer type [enabled by default]
C:/PROGRA~1/R/R-31~1.2/include/iconv.h:53:8: note: expected 'const char **' but argument is of type 'char **'
src/readstat_sav_parse.rl:109:17: warning: format '%ld' expects argument of type 'long int', but argument 4 has type 'long long int' [-Wformat]
src/readstat_sav_parse.rl: In function 'sav_parse_very_long_string_record':
src/readstat_sav_parse.rl:154:17: warning: passing argument 2 of 'libiconv' from incompatible pointer type [enabled by default]
C:/PROGRA~1/R/R-31~1.2/include/iconv.h:53:8: note: expected 'const char **' but argument is of type 'char **'
src/readstat_sav_parse.rl:204:9: warning: format '%ld' expects argument of type 'long int', but argument 3 has type 'long long int' [-Wformat]
src/readstat_sav_parse.rl:204:9: warning: format '%ld' expects argument of type 'long int', but argument 4 has type 'long long int' [-Wformat]
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"C:/Users/Jeroen/Documents/R/win-library/3.1/BH/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"     -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_spss.c -o readstat_spss.o
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"C:/Users/Jeroen/Documents/R/win-library/3.1/BH/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"     -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_value.c -o readstat_value.o
g++ -m64 -shared -s -static-libgcc -o haven.dll tmp.def CKHashTable.o DfBuilder.o RcppExports.o readstat_bits.o readstat_convert.o readstat_dta.o readstat_error.o readstat_io.o readstat_por.o readstat_por_parse.o readstat_sas.o readstat_sav.o readstat_sav_parse.o readstat_spss.o readstat_value.o -lRiconv -Ld:/RCompile/CRANpkg/extralibs64/local/lib/x64 -Ld:/RCompile/CRANpkg/extralibs64/local/lib -LC:/PROGRA~1/R/R-31~1.2/bin/x64 -lR
installing to C:/Users/Jeroen/Documents/R/win-library/3.1/haven/libs/x64
** R
** preparing package for lazy loading
** help
*** installing help indices
** building package indices
** testing if installed package can be loaded
* DONE (haven)
```

**Comment by evanmiller** [comment]
This commit should fix all of the non-iconv warnings:

https://github.com/WizardMac/ReadStat/commit/b4bfe775835176bbbd1c27e65a269f7b06ce2398

**Comment by evanmiller** [comment]
Finally this commit only includes the LZMA/XZ stuff if HAVE_LZMA is defined:

https://github.com/WizardMac/ReadStat/commit/11349025ca996b870281fb33dd4f65e2f7632c7d

So hopefully the library will now compile on Windows without having to manually remove the RData code.

**Comment by jeroenooms** [comment]
Almost there. A few warnings remain:

```c
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"C:/Users/Jeroen/Documents/R/win-library/3.1/BH/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"     -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_dta.c -o readstat_dta.o
readstat_convert.c: In function 'readstat_convert':
readstat_convert.c:23:9: warning: passing argument 2 of 'libiconv' from incompatible pointer type [enabled by default]
C:/PROGRA~1/R/R-31~1.2/include/iconv.h:53:8: note: expected 'const char **' but argument is of type 'char **'
```

```c
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"C:/Users/Jeroen/Documents/R/win-library/3.1/BH/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"     -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_rdata.c -o readstat_rdata.o
src/readstat_por_parse.rl: In function 'readstat_por_parse_double':
src/readstat_por_parse.rl:69:9: warning: format '%ld' expects argument of type 'long int', but argument 3 has type 'long long int' [-Wformat]
```

```c
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"C:/Users/Jeroen/Documents/R/win-library/3.1/BH/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"     -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_sav_parse.c -o readstat_sav_parse.o
src/readstat_sav_parse.rl: In function 'sav_parse_long_variable_names_record':
src/readstat_sav_parse.rl:56:17: warning: passing argument 2 of 'libiconv' from incompatible pointer type [enabled by default]
C:/PROGRA~1/R/R-31~1.2/include/iconv.h:53:8: note: expected 'const char **' but argument is of type 'char **'
src/readstat_sav_parse.rl: In function 'sav_parse_very_long_string_record':
src/readstat_sav_parse.rl:154:17: warning: passing argument 2 of 'libiconv' from incompatible pointer type [enabled by default]
C:/PROGRA~1/R/R-31~1.2/include/iconv.h:53:8: note: expected 'const char **' but argument is of type 'char **'
```

**Comment by evanmiller** [comment]
This commit should fix the second warning:

https://github.com/WizardMac/ReadStat/commit/f7971ff9e00b66ba79e23be69b262b1c6126a2bf

As for the iconv warnings -- it looks like the API signature is non-standard. Apparently R uses win_iconv on Windows and for some reason win_iconv added the `const` keyword to `inbuf`. I'm not sure what the proper resolution is since casting to `(char **)` will break the code on Windows and `(const char **)` will break the code on Unix.

**Comment by jeroenooms** [comment]
How about 

```c
#ifdef WIN32
..
#else
..
#endif
```

**Comment by evanmiller** [comment]
Ok try this commit then:

https://github.com/WizardMac/ReadStat/commit/b871c8e0ea4d3ef02a6e4dcb4bda1381de3928dc

**Comment by jeroenooms** [comment]
OK no more warnings in Windows with the standard compiler flags. Yay, great job! 

CRAN sometimes complains about warnings from the -Wall -pedantic` flags. These are not super important but if you can fix them as well that would be awesome:

```c
==> Rcmd.exe INSTALL --no-multiarch --with-keep.source haven

* installing to library 'C:/Users/Jeroen/Documents/R/win-library/3.1'
* installing *source* package 'haven' ...
** libs
rm -f CKHashTable.o readstat_bits.o readstat_convert.o readstat_dta.o readstat_error.o readstat_io.o readstat_por.o readstat_por_parse.o readstat_rdata.o readstat_sas.o readstat_sav.o readstat_sav_parse.o readstat_spss.o readstat_value.o
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"C:/Users/Jeroen/Documents/R/win-library/3.1/BH/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"  -Wall -pedantic   -O2 -Wall  -std=gnu99 -mtune=core2 -c CKHashTable.c -o CKHashTable.o
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"C:/Users/Jeroen/Documents/R/win-library/3.1/BH/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"  -Wall -pedantic   -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_bits.c -o readstat_bits.o
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"C:/Users/Jeroen/Documents/R/win-library/3.1/BH/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"  -Wall -pedantic   -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_convert.c -o readstat_convert.o
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"C:/Users/Jeroen/Documents/R/win-library/3.1/BH/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"  -Wall -pedantic   -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_dta.c -o readstat_dta.o
In file included from readstat_dta.c:9:0:
readstat_dta.h:20:22: warning: ISO C forbids zero-size array 'contents' [-pedantic]
readstat_dta.h:26:22: warning: ISO C forbids zero-size array 'contents' [-pedantic]
readstat_dta.c: In function 'dta_ctx_init':
readstat_dta.c:31:9: warning: overflow in implicit constant conversion [-Woverflow]
readstat_dta.c:92:9: warning: overflow in implicit constant conversion [-Woverflow]
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"C:/Users/Jeroen/Documents/R/win-library/3.1/BH/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"  -Wall -pedantic   -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_error.c -o readstat_error.o
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"C:/Users/Jeroen/Documents/R/win-library/3.1/BH/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"  -Wall -pedantic   -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_io.c -o readstat_io.o
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"C:/Users/Jeroen/Documents/R/win-library/3.1/BH/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"  -Wall -pedantic   -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_por.c -o readstat_por.o
readstat_por.c: In function 'read_bytes':
readstat_por.c:140:20: warning: pointer of type 'void *' used in arithmetic [-pedantic]
readstat_por.c:166:20: warning: pointer of type 'void *' used in arithmetic [-pedantic]
readstat_por.c: In function 'utf8_encode':
readstat_por.c:300:13: warning: ISO C does not support the '%C' ms_printf format [-Wformat]
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"C:/Users/Jeroen/Documents/R/win-library/3.1/BH/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"  -Wall -pedantic   -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_por_parse.c -o readstat_por_parse.o
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"C:/Users/Jeroen/Documents/R/win-library/3.1/BH/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"  -Wall -pedantic   -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_rdata.c -o readstat_rdata.o
readstat_rdata.c: In function 'read_st_z':
readstat_rdata.c:87:40: warning: pointer of type 'void *' used in arithmetic [-pedantic]
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"C:/Users/Jeroen/Documents/R/win-library/3.1/BH/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"  -Wall -pedantic   -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_sas.c -o readstat_sas.o
readstat_sas.c:53:5: warning: overflow in implicit constant conversion [-Woverflow]
readstat_sas.c:53:5: warning: overflow in implicit constant conversion [-Woverflow]
readstat_sas.c:53:5: warning: overflow in implicit constant conversion [-Woverflow]
readstat_sas.c:54:5: warning: overflow in implicit constant conversion [-Woverflow]
readstat_sas.c:54:5: warning: overflow in implicit constant conversion [-Woverflow]
readstat_sas.c:54:5: warning: overflow in implicit constant conversion [-Woverflow]
readstat_sas.c:54:5: warning: overflow in implicit constant conversion [-Woverflow]
readstat_sas.c:55:5: warning: overflow in implicit constant conversion [-Woverflow]
readstat_sas.c:55:5: warning: overflow in implicit constant conversion [-Woverflow]
readstat_sas.c:60:5: warning: overflow in implicit constant conversion [-Woverflow]
readstat_sas.c:60:5: warning: overflow in implicit constant conversion [-Woverflow]
readstat_sas.c:60:5: warning: overflow in implicit constant conversion [-Woverflow]
readstat_sas.c:61:5: warning: overflow in implicit constant conversion [-Woverflow]
readstat_sas.c:61:5: warning: overflow in implicit constant conversion [-Woverflow]
readstat_sas.c:61:5: warning: overflow in implicit constant conversion [-Woverflow]
readstat_sas.c:61:5: warning: overflow in implicit constant conversion [-Woverflow]
readstat_sas.c:62:5: warning: overflow in implicit constant conversion [-Woverflow]
readstat_sas.c:62:5: warning: overflow in implicit constant conversion [-Woverflow]
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"C:/Users/Jeroen/Documents/R/win-library/3.1/BH/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"  -Wall -pedantic   -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_sav.c -o readstat_sav.o
readstat_sav.c: In function 'sav_ctx_init':
readstat_sav.c:140:9: warning: overflow in implicit constant conversion [-Woverflow]
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"C:/Users/Jeroen/Documents/R/win-library/3.1/BH/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"  -Wall -pedantic   -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_sav_parse.c -o readstat_sav_parse.o
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"C:/Users/Jeroen/Documents/R/win-library/3.1/BH/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"  -Wall -pedantic   -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_spss.c -o readstat_spss.o
gcc -m64 -I"C:/PROGRA~1/R/R-31~1.2/include" -DNDEBUG    -I"C:/Users/Jeroen/Documents/R/win-library/3.1/Rcpp/include" -I"C:/Users/Jeroen/Documents/R/win-library/3.1/BH/include" -I"d:/RCompile/CRANpkg/extralibs64/local/include"  -Wall -pedantic   -O2 -Wall  -std=gnu99 -mtune=core2 -c readstat_value.c -o readstat_value.o
g++ -m64 -shared -s -static-libgcc -o haven.dll tmp.def CKHashTable.o readstat_bits.o readstat_convert.o readstat_dta.o readstat_error.o readstat_io.o readstat_por.o readstat_por_parse.o readstat_rdata.o readstat_sas.o readstat_sav.o readstat_sav_parse.o readstat_spss.o readstat_value.o -lRzlib -lRiconv -Ld:/RCompile/CRANpkg/extralibs64/local/lib/x64 -Ld:/RCompile/CRANpkg/extralibs64/local/lib -LC:/PROGRA~1/R/R-31~1.2/bin/x64 -lR
installing to C:/Users/Jeroen/Documents/R/win-library/3.1/haven/libs/x64
** R
** preparing package for lazy loading
** help
*** installing help indices
** building package indices
** testing if installed package can be loaded
* DONE (haven)
```

**Comment by evanmiller** [comment]
Ok, I think I've fixed all the pedantic warnings. Grab the latest code and see if it compiles cleanly.

**Comment by hadley** [comment]
Looks good to me :)

**Comment by evanmiller** [comment]
Closing for now. If new warnings arise on any platform please open a new issue.

**Comment by diogocp** [comment]
New warnings in #28.
